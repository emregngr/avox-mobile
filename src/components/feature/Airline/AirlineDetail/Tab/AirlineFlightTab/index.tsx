import React, { useMemo } from 'react'
import { View } from 'react-native'

import { Alliance } from '@/components/feature/Airline/AirlineDetail/Tab/AirlineFlightTab/Sections/Alliance'
import { DestinationList } from '@/components/feature/Airline/AirlineDetail/Tab/AirlineFlightTab/Sections/DestinationList'
import { RoutesList } from '@/components/feature/Airline/AirlineDetail/Tab/AirlineFlightTab/Sections/RoutesList'
import { StatsGrid } from '@/components/feature/Airline/AirlineDetail/Tab/AirlineFlightTab/Sections/StatsGrid'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { Airline } from '@/types/feature/airline'

interface AirlineFlightTabProps {
  airlineData: Airline
}

export const AirlineFlightTab = ({ airlineData }: AirlineFlightTabProps) => {
  const { selectedTheme } = useThemeStore()
  const { selectedLocale } = useLocaleStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { network, operations } = airlineData ?? {}
  const { alliance } = operations ?? {}
  const {
    destinationCount,
    destinationCountries,
    destinations,
    domesticConnections,
    internationalConnections,
    routes,
  } = network ?? {}

  const memoizedStats = useMemo(
    () => ({
      destinationCount,
      destinationCountries,
      domesticConnections,
      internationalConnections,
    }),
    [destinationCount, destinationCountries, domesticConnections, internationalConnections],
  )

  const iconColor = useMemo(() => colors.onPrimary100, [colors.onPrimary100])

  const localeStrings = useMemo(
    () => ({
      activeRoutes: getLocale('activeRoutes'),
      alliance: getLocale('alliance'),
      country: getLocale('country'),
      domesticDestinations: getLocale('domesticDestinations'),
      internationalDestination: getLocale('internationalDestination'),
      mainDestinations: getLocale('mainDestinations'),
      totalDestination: getLocale('totalDestination'),
    }),
    [selectedLocale],
  )

  return (
    <View className="px-4 gap-y-4">
      <StatsGrid
        localeStrings={{
          country: localeStrings.country,
          domesticDestinations: localeStrings.domesticDestinations,
          internationalDestination: localeStrings.internationalDestination,
          totalDestination: localeStrings.totalDestination,
        }}
        iconColor={iconColor}
        stats={memoizedStats}
      />

      <DestinationList
        destinations={destinations}
        iconColor={iconColor}
        title={localeStrings.mainDestinations}
      />

      <RoutesList iconColor={iconColor} routes={routes} title={localeStrings.activeRoutes} />

      {alliance !== 'none' && (
        <Alliance alliance={alliance} iconColor={iconColor} title={localeStrings.alliance} />
      )}
    </View>
  )
}
