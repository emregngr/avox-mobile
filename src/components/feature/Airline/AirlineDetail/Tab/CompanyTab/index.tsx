import React from 'react'
import { View } from 'react-native'

import { Company } from '@/components/feature/Airline/AirlineDetail/Tab/CompanyTab/Sections/Company'
import { Contact } from '@/components/feature/Airline/AirlineDetail/Tab/CompanyTab/Sections/Contact'
import { Hub } from '@/components/feature/Airline/AirlineDetail/Tab/CompanyTab/Sections/Hub'
import { Map } from '@/components/feature/Airline/AirlineDetail/Tab/CompanyTab/Sections/Map'
import { SocialMedia } from '@/components/feature/Airline/AirlineDetail/Tab/CompanyTab/Sections/SocialMedia'
import type { Airline } from '@/types/feature/airline'

interface CompanyTabProps {
  airlineData: Airline
}

export const CompanyTab = ({ airlineData }: CompanyTabProps) => {
  const { companyInfo, operations } = airlineData

  return (
    <View className="px-4 gap-y-4">
      <Company airlineData={airlineData} />

      <Contact companyInfo={companyInfo} operations={operations} />

      <SocialMedia companyInfo={companyInfo} />

      <Hub operations={operations} />

      <Map airlineData={airlineData} />
    </View>
  )
}
