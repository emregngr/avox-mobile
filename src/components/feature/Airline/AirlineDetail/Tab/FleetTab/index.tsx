import { router } from 'expo-router'
import React, { useCallback, useMemo } from 'react'
import { View } from 'react-native'

import { FleetHeader } from '@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Sections/FleetHeader'
import { FleetList } from '@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Sections/FleetList'
import { FleetStats } from '@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Sections/FleetStats'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { Airline } from '@/types/feature/airline'

interface FleetTabProps {
  airlineData: Airline
}

export const FleetTab = ({ airlineData }: FleetTabProps) => {
  const { selectedTheme } = useThemeStore()
  const { selectedLocale } = useLocaleStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { fleet, operations } = airlineData || {}

  const { region } = operations || {}

  const { airplaneTypeCount, airplanes, averageAgeYears, totalAirplane } = fleet || {}

  const sortedAirplanes = useMemo(
    () => airplanes?.sort((a, b) => b.count - a.count) || [],
    [airplanes],
  )

  const regionLower = useMemo(() => region?.toLowerCase(), [region])

  const handleImagePress = useCallback(
    (airplaneType: string, imageKey: string) => {
      router.navigate({
        params: {
          regionLower,
          selectedImageKey: imageKey,
          title: airplaneType,
        },
        pathname: '/image-modal',
      })
    },
    [regionLower],
  )

  const localeStrings = useMemo(
    () => ({
      fleetDetail: getLocale('fleetDetail'),
    }),
    [selectedLocale],
  )

  return (
    <View className="px-4 gap-y-4">
      <FleetStats
        airplaneTypeCount={airplaneTypeCount}
        averageAgeYears={averageAgeYears}
        totalAirplane={totalAirplane}
      />

      <FleetHeader colors={colors} fleetDetailText={localeStrings.fleetDetail} />

      <FleetList
        airplanes={sortedAirplanes}
        colors={colors}
        onImagePress={handleImagePress}
        region={region}
        totalAirplane={totalAirplane}
      />
    </View>
  )
}
