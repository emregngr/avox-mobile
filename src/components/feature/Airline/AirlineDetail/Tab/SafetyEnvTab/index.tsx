import React, { useMemo } from 'react'
import { View } from 'react-native'

import { Environmental } from '@/components/feature/Airline/AirlineDetail/Tab/SafetyEnvTab/Sections/Environmental'
import { Safety } from '@/components/feature/Airline/AirlineDetail/Tab/SafetyEnvTab/Sections/Safety'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { AirlineType } from '@/types/feature/airline'

interface SafetyEnvTabProps {
  airlineData: AirlineType
}

export const SafetyEnvTab = ({ airlineData }: SafetyEnvTabProps) => {
  const { selectedLocale } = useLocaleStore()
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { environmental, safety } = airlineData ?? {}
  const { certifications, safetyRecord } = safety ?? {}

  const localeStrings = useMemo(
    () => ({
      certifications: getLocale('certifications'),
      environmentalPolicy: getLocale('environmentalPolicy'),
      environmentalResponsibility: getLocale('environmentalResponsibility'),
      safetyInformation: getLocale('safetyInformation'),
      safetyRecord: getLocale('safetyRecord'),
    }),
    [selectedLocale],
  )

  return (
    <View className="px-4 gap-y-4">
      <Safety
        certifications={certifications}
        certificationsTitle={localeStrings.certifications}
        iconColor={colors.onPrimary100}
        safetyRecord={safetyRecord}
        safetyRecordTitle={localeStrings.safetyRecord}
        title={localeStrings.safetyInformation}
      />

      <Environmental
        content={environmental}
        iconColor={colors.onPrimary100}
        subtitle={localeStrings.environmentalResponsibility}
        title={localeStrings.environmentalPolicy}
      />
    </View>
  )
}
