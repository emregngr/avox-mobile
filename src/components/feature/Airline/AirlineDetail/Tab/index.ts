export * from './AirlineFlightTab'
export * from './CompanyTab'
export * from './FleetTab'
export * from './SafetyEnvTab'
