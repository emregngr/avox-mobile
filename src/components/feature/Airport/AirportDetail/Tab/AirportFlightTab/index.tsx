import { useMemo } from 'react'
import { View } from 'react-native'

import { AirlinesList } from '@/components/feature/Airport/AirportDetail/Tab/AirportFlightTab/Sections/AirlinesList'
import { RoutesList } from '@/components/feature/Airport/AirportDetail/Tab/AirportFlightTab/Sections/RoutesList'
import { StatsGrid } from '@/components/feature/Airport/AirportDetail/Tab/AirportFlightTab/Sections/StatsGrid'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { AirportType } from '@/types/feature/airport'

interface AirportFlightTabProps {
  airportData: AirportType
}

export const AirportFlightTab = ({ airportData }: AirportFlightTabProps) => {
  const { selectedLocale } = useLocaleStore()
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { flightOperations } = airportData ?? {}
  const {
    airlines,
    destinationCount,
    destinationCountries,
    domesticConnections,
    internationalConnections,
    routes,
  } = flightOperations ?? {}

  const memoizedStats = useMemo(
    () => ({
      destinationCount,
      destinationCountries,
      domesticConnections,
      internationalConnections,
    }),
    [destinationCount, destinationCountries, domesticConnections, internationalConnections],
  )

  const localeStrings = useMemo(
    () => ({
      airlines: getLocale('airlines'),
      country: getLocale('country'),
      domesticDestinations: getLocale('domesticDestinations'),
      internationalDestination: getLocale('internationalDestination'),
      popularDestinations: getLocale('popularDestinations'),
      totalDestination: getLocale('totalDestination'),
    }),
    [selectedLocale],
  )

  return (
    <View className="px-4 gap-y-4">
      <StatsGrid
        localeStrings={{
          country: localeStrings.country,
          domesticDestinations: localeStrings.domesticDestinations,
          internationalDestination: localeStrings.internationalDestination,
          totalDestination: localeStrings.totalDestination,
        }}
        iconColor={colors?.onPrimary100}
        stats={memoizedStats}
      />

      <AirlinesList
        airlines={airlines}
        iconColor={colors?.onPrimary100}
        title={localeStrings.airlines}
      />

      <RoutesList
        iconColor={colors?.onPrimary100}
        routes={routes}
        title={localeStrings.popularDestinations}
      />
    </View>
  )
}
