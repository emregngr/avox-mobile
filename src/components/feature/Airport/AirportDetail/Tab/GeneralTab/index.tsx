import React from 'react'
import { View } from 'react-native'

import { Contact } from '@/components/feature/Airport/AirportDetail/Tab/GeneralTab/Sections/Contact'
import { General } from '@/components/feature/Airport/AirportDetail/Tab/GeneralTab/Sections/General'
import { Location } from '@/components/feature/Airport/AirportDetail/Tab/GeneralTab/Sections/Location'
import { Map } from '@/components/feature/Airport/AirportDetail/Tab/GeneralTab/Sections/Map'
import { Services } from '@/components/feature/Airport/AirportDetail/Tab/GeneralTab/Sections/Services'
import { SocialMedia } from '@/components/feature/Airport/AirportDetail/Tab/GeneralTab/Sections/SocialMedia'
import type { Airport } from '@/types/feature/airport'

interface GeneralTabProps {
  airportData: Airport
}

export const GeneralTab = ({ airportData }: GeneralTabProps) => {
  const { airportInfo, facilities, operations } = airportData ?? {}

  return (
    <View className="px-4 gap-y-4">
      <General airportInfo={airportInfo} operations={operations} />

      <Services facilities={facilities} />

      <Contact airportInfo={airportInfo} />

      <SocialMedia airportInfo={airportInfo} />

      <Location operations={operations} />

      <Map airportData={airportData} />
    </View>
  )
}
