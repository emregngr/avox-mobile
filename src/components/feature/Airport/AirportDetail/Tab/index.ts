export * from './AirportFlightTab'
export * from './GeneralTab'
export * from './InfrastructureTab'
export * from './NearbyPlacesTab'
