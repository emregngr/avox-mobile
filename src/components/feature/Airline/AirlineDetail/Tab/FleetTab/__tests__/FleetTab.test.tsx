import { act, render } from '@testing-library/react-native'
import { router } from 'expo-router'
import React from 'react'

import { FleetTab } from '@/components/feature/Airline/AirlineDetail/Tab/FleetTab'
import { getLocale } from '@/locales/i18next'
import useLocaleStore from '@/store/locale'

jest.mock('@/locales/i18next')

const mockedGetLocale = getLocale as jest.MockedFunction<typeof getLocale>

jest.mock('@/store/locale')

const mockedUseLocaleStore = useLocaleStore as jest.MockedFunction<typeof useLocaleStore>

const mockedFleetStats = jest.fn()
const mockedFleetHeader = jest.fn()
const mockedFleetList = jest.fn()

jest.mock('@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Sections/FleetStats', () => {
  const { View } = require('react-native')
  return {
    FleetStats: (props: any) => {
      mockedFleetStats(props)
      return <View testID="mocked-fleet-stats" {...props} />
    },
  }
})

jest.mock('@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Sections/FleetHeader', () => {
  const { View } = require('react-native')
  return {
    FleetHeader: (props: any) => {
      mockedFleetHeader(props)
      return <View testID="mocked-fleet-header" {...props} />
    },
  }
})

jest.mock('@/components/feature/Airline/AirlineDetail/Tab/FleetTab/Sections/FleetList', () => {
  const { View } = require('react-native')
  return {
    FleetList: (props: any) => {
      mockedFleetList(props)
      return <View testID="mocked-fleet-list" {...props} />
    },
  }
})

const mockedAirlineData: any = {
  fleet: {
    airplanes: [
      { airplaneType: 'Airbus A320', count: 15, imageKeys: [] },
      { airplaneType: 'Boeing 777', count: 25, imageKeys: [] },
      { airplaneType: 'Boeing 737', count: 20, imageKeys: [] },
    ],
    airplaneTypeCount: 3,
    averageAgeYears: 12.5,
    totalAirplane: 60,
  },
  operations: {
    region: 'North America',
  },
}

beforeEach(() => {
  mockedUseLocaleStore.mockReturnValue({ selectedLocale: 'en' })

  mockedGetLocale.mockReturnValue('Mocked Fleet Details')
})

describe('FleetTab Component', () => {
  it('should render all child components', () => {
    const { getByTestId } = render(<FleetTab airlineData={mockedAirlineData} />)

    expect(getByTestId('mocked-fleet-stats')).toBeTruthy()
    expect(getByTestId('mocked-fleet-header')).toBeTruthy()
    expect(getByTestId('mocked-fleet-list')).toBeTruthy()
  })

  describe('Prop Passing and Component Calls', () => {
    it('should pass correct props to FleetStats', () => {
      render(<FleetTab airlineData={mockedAirlineData} />)

      expect(mockedFleetStats).toHaveBeenCalledWith(
        expect.objectContaining({
          airplaneTypeCount: mockedAirlineData.fleet.airplaneTypeCount,
          averageAgeYears: mockedAirlineData.fleet.averageAgeYears,
          totalAirplane: mockedAirlineData.fleet.totalAirplane,
        }),
      )
    })

    it('should pass correct props to FleetHeader', () => {
      render(<FleetTab airlineData={mockedAirlineData} />)

      expect(mockedGetLocale).toHaveBeenCalledWith('fleetDetail')
      expect(mockedFleetHeader).toHaveBeenCalledWith(
        expect.objectContaining({
          fleetDetailText: 'Mocked Fleet Details',
        }),
      )
    })

    it('should sort airplanes and pass correct props to FleetList', () => {
      render(<FleetTab airlineData={mockedAirlineData} />)

      const sortedAirplanes = [
        { airplaneType: 'Boeing 777', count: 25, imageKeys: [] },
        { airplaneType: 'Boeing 737', count: 20, imageKeys: [] },
        { airplaneType: 'Airbus A320', count: 15, imageKeys: [] },
      ]
      expect(mockedFleetList).toHaveBeenCalledWith(
        expect.objectContaining({
          airplanes: sortedAirplanes,
          region: mockedAirlineData.operations.region,
          totalAirplane: mockedAirlineData.fleet.totalAirplane,
          onImagePress: expect.any(Function),
        }),
      )
    })

    it('should call all components exactly once', () => {
      render(<FleetTab airlineData={mockedAirlineData} />)

      expect(mockedFleetStats).toHaveBeenCalledTimes(1)
      expect(mockedFleetHeader).toHaveBeenCalledTimes(1)
      expect(mockedFleetList).toHaveBeenCalledTimes(1)
    })
  })

  describe('Functionality and Edge Cases', () => {
    it('should call router.navigate with correct parameters when onImagePress is triggered', () => {
      render(<FleetTab airlineData={mockedAirlineData} />)

      const onImagePressCallback = mockedFleetList.mock.calls[0][0].onImagePress
      const testAirplaneType = 'Test Plane'
      const testImageKey = 'test-image-123'
      act(() => {
        onImagePressCallback(testAirplaneType, testImageKey)
      })
      expect(router.navigate).toHaveBeenCalledTimes(1)
      expect(router.navigate).toHaveBeenCalledWith({
        pathname: '/image-modal',
        params: {
          title: testAirplaneType,
          selectedImageKey: testImageKey,
        },
      })
    })

    it('should handle partially incomplete data gracefully', () => {
      const incompleteData: any = {
        operations: {
          region: 'Asia',
        },
      }
      render(<FleetTab airlineData={incompleteData} />)

      expect(mockedFleetStats).toHaveBeenCalledWith(
        expect.objectContaining({
          airplaneTypeCount: undefined,
          averageAgeYears: undefined,
          totalAirplane: undefined,
        }),
      )
      expect(mockedFleetList).toHaveBeenCalledWith(
        expect.objectContaining({
          airplanes: undefined,
          region: 'Asia',
          totalAirplane: undefined,
        }),
      )
    })

    it('should handle completely empty data without crashing', () => {
      const { getByTestId } = render(<FleetTab airlineData={{} as any} />)

      expect(getByTestId('mocked-fleet-stats')).toBeTruthy()
      expect(getByTestId('mocked-fleet-header')).toBeTruthy()
      expect(getByTestId('mocked-fleet-list')).toBeTruthy()
      expect(mockedFleetList).toHaveBeenCalledWith(
        expect.objectContaining({
          airplanes: undefined,
          region: undefined,
        }),
      )
    })
  })
})

describe('FleetTab Component Snapshot', () => {
  it('should render the FleetTab Component successfully', () => {
    const { toJSON } = render(<FleetTab airlineData={mockedAirlineData} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
