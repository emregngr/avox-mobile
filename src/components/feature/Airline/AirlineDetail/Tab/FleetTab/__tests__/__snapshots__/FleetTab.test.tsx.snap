// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FleetTab Component Snapshot should render the FleetTab Component successfully 1`] = `
<View
  className="px-4 gap-y-4"
>
  <View
    airplaneTypeCount={3}
    averageAgeYears={12.5}
    testID="mocked-fleet-stats"
    totalAirplane={60}
  />
  <View
    fleetDetailText="Mocked Fleet Details"
    testID="mocked-fleet-header"
  />
  <View
    airplanes={
      [
        {
          "airplaneType": "Boeing 777",
          "count": 25,
          "imageKeys": [],
        },
        {
          "airplaneType": "Boeing 737",
          "count": 20,
          "imageKeys": [],
        },
        {
          "airplaneType": "Airbus A320",
          "count": 15,
          "imageKeys": [],
        },
      ]
    }
    onImagePress={[Function]}
    region="North America"
    testID="mocked-fleet-list"
    totalAirplane={60}
  />
</View>
`;
